configure.py

```diff
@@ -1115,7 +1118,7 @@ def query_seastar_flags(seastar_pc_file, link_static_cxx=False):
 args.user_cflags += ' -march=' + args.target
 libs = ' '.join([maybe_static(args.staticyamlcpp, '-lyaml-cpp'), '-latomic', '-llz4', '-lz', '-lsnappy', pkg_config('jsoncpp', '--libs'),
                  ' -lstdc++fs', ' -lcrypt', ' -lcryptopp', ' -lpthread',
-                 maybe_static(args.staticboost, '-lboost_date_time'), ])
```

Comment:
Doesn't `std::regex` cut it?

Reply:
No.  I tried it three different ways, all weird and painful.  I'm happy to discuss the details if desired, but the bottom line is that Boost (with ICU) is the best choice.

Reply:
I believe you. :smile: The standard library is still *very* awkward to use with anything more than plain old ASCII.

Reply:
I'm interested in the details.

Reply:
First I tried using explicit-width characters (such as `char32_t`), because there are locale-independent library functions to convert UTF-8 into them.  But it turns out you [can't have](https://stackoverflow.com/a/33709665) `std::regex` with such characters OOTB.  You'd have to explicitly implement `std::regex_traits<char32_t>`, which is fraught with pitfalls and would take forever.

Then I tried using `wchar_t`, but there are no explicit library functions for converting UTF-8 to `wchar_t`:
- `std::codecvt_utf8` is deprecated
- `mbrtowc` depends on the globally set locale in an underdocumented way that doesn't explain thread safety, how to switch the global locale back to whatever it was before, or why one needs to specify a language just to convert UTF-8 to `wchar_t`
- `std::codecvt` is only locale-independent for explicit-width characters and has a protected destructor

Reply:
@dekimir @avikivity I'm trying to understand better why a like matcher needs to understand UTF-8, and can't just treat a single UTF-8 code point as multiple bytes.
Maybe the only reason it needs to understand UTF-8 is to support "_" as a way to match one and only one character (possibly multiple bytes in utf-8)?
If this is the **only** reason, it would be fairly trivial to write a regex to match one UTF-8 character (one start byte plus zero or more continuation bytes). No need to bring the entire ICU and Boost into our code just for this.
So, is there another reason?

Reply:
First off, keep in mind that with Unicode encodings, there is _always_ some case you missed, and your homegrown solution will fail.  Virtually everyone who ever tried a homegrown solution concluded in the end that it's cheaper and easier to use ICU from the get-go.

> it would be fairly trivial to write a regex to match one UTF-8 character (one start byte plus zero or more continuation bytes)

You have to interpret the bits of that first byte, no?  And aren't there multiple encodings for some characters (ie, different byte sequences that represent the same character and must therefore match each other)?  I wouldn't call it trivial, just possible.  It would, of course, amount to a homegrown solution -- see the first paragraph. :)

Reply:
Since this is an old closed PR, I moved my proposal to a new issue: https://github.com/scylladb/scylla/issues/7843

I don't think that a "homegrown" solution is bad when the problem is trivial. ICU can do a lot of fancy stuff on UTF-8, none of which you need in your case. You don't even need character classes (like `[[:alpha:]]`) which ICU's huge tables are all about.

Your last question is interesting, but luckily (or rather, intentionally), the answer is no: official (validated) UTF-8 does not allow two different encodings of the same character - one encoding is considered correct, and the others are considered invalid. There are cases when two sequences of one or more utf-8 characters *look* the same, but there is no reason to consider those "equal".